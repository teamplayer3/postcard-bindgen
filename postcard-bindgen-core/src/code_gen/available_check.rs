use super::variable_path::{VariableAccess, VariablePath};

#[derive(Clone)]
pub enum AvailableCheck<L> {
    Object(VariablePath<L>, String),
    None,
}

impl<L> AvailableCheck<L> {
    pub fn from_variable_path(path: VariablePath<L>) -> Self {
        let (path, last) = path.modify_pop();
        match last {
            Some(VariableAccess::Field(name)) => Self::Object(path, name),
            _ => Self::None,
        }
    }
}
