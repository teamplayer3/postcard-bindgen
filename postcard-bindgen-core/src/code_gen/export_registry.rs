use genco::{lang::Lang, quote, tokens::FormatInto, Tokens};

#[derive(Debug, Default, Clone)]
pub enum ExportMode {
    #[default]
    Cjs,
    Esm,
}

pub struct ExportRegistry<L>
where
    L: Lang,
{
    pub(super) exports: Vec<Tokens<L>>,
    pub(super) export_mode: ExportMode,
}

impl<L> ExportRegistry<L>
where
    L: Lang,
{
    pub fn new(export_mode: ExportMode) -> Self {
        Self {
            exports: Vec::new(),
            export_mode,
        }
    }
}

impl<L> ExportRegistry<L>
where
    L: Lang,
{
    pub fn push(&mut self, export: impl FormatInto<L>) {
        self.exports.push(quote!($export));
    }
}
