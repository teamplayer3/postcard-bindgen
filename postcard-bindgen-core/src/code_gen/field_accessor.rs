#[derive(Debug, Clone, Copy)]
pub enum FieldAccessor<'a> {
    Object(&'a str),
    Array,
    None,
}
