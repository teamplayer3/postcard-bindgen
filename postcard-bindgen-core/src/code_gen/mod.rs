mod utils;

#[cfg(feature = "js")]
mod js;

#[cfg(feature = "python")]
mod python;

#[cfg(feature = "js")]
pub use js::{generate_js as generate, type_checking::ts::gen_ts_typings as generate_typings};

#[cfg(feature = "python")]
pub use python::{generate, type_checking::typing::generate_typings};