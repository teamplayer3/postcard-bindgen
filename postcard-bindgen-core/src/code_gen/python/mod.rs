mod des;
mod general;
mod generateable;
mod ser;
mod type_checks;

use core::borrow::Borrow;

use convert_case::{Case, Casing};
use des::{gen_des_functions, gen_deserialize_func, gen_deserializer_code};
use genco::{lang::python::Python, quote, quote_in, tokens::FormatInto};
use general::gen_util;
use generateable::{gen_basic_typings, gen_typings};
use ser::{gen_ser_functions, gen_serialize_func, gen_serializer_code};
use type_checks::gen_type_checks;

use crate::{
    code_gen::import_registry::ImportMode, path::PathBuf, registry::ContainerCollection, Exports,
};

use super::{
    import_registry::{ImportItem, Package},
    utils::{IfBranchedTemplate, TokensBranchedIterExt, TokensIterExt},
};

const PYTHON_OBJECT_VARIABLE: &str = "v";
const PYTHON_LOGIC_AND: &str = "and";
const PYTHON_LOGIC_OR: &str = "or";

type Tokens = genco::lang::python::Tokens;

type VariablePath = super::variable_path::VariablePath<Python>;
type VariableAccess = super::variable_path::VariableAccess;
type FieldAccessor<'a> = super::field_accessor::FieldAccessor<'a>;
type AvailableCheck = super::available_check::AvailableCheck<Python>;
type ImportRegistry = super::import_registry::ImportRegistry;
type ExportFile = crate::ExportFile<Python>;

/// Settings for bindings generation.
///
/// This enables the possibility to enable or disable serialization, deserialization, runtime type checks
/// or type script types.
/// Less code will be generated if an option is off.
///
/// By default, only deserialization is enabled. Serialization can be enabled by using [`GenerationSettings::serialization()`].
/// Deserialization can be disabled with [`GenerationSettings::deserialization()`].
/// To enable all at once use [`GenerationSettings::enable_all()`].
#[derive(Debug)]
pub struct GenerationSettings {
    ser: bool,
    des: bool,
    runtime_type_checks: bool,
    module_structure: bool,
}

impl GenerationSettings {
    /// Constructs [`GenerationSettings`] and enables all options at once.
    pub fn enable_all() -> Self {
        Self {
            ser: true,
            des: true,
            runtime_type_checks: true,
            module_structure: true,
        }
    }

    /// Enabling or disabling of serialization code generation.
    pub fn serialization(mut self, enabled: bool) -> Self {
        self.ser = enabled;
        self
    }

    /// Enabling or disabling of deserialization code generation.
    pub fn deserialization(mut self, enabled: bool) -> Self {
        self.des = enabled;
        self
    }

    /// Enabling or disabling of runtime type checks code generation.
    ///
    /// Disabling this should lead to a speed increase at serialization.
    pub fn runtime_type_checks(mut self, enabled: bool) -> Self {
        self.runtime_type_checks = enabled;
        self
    }

    /// Enabling or disabling of module structure code generation.
    ///
    /// Enabling this will generate the types in the same module structure
    /// as in rust. Root level types will be in the root of the generated
    /// module (subpackage types). Types nested in modules will be in subpackages
    /// (e.g. types.<mod_name>.<type_name>). This avoids name clashes.
    ///
    /// Disabling this will generate all types in the root module.
    pub fn module_structure(mut self, enabled: bool) -> Self {
        self.module_structure = enabled;
        self
    }
}

impl Default for GenerationSettings {
    fn default() -> Self {
        Self {
            ser: false,
            des: true,
            runtime_type_checks: false,
            module_structure: true,
        }
    }
}

pub fn generate(
    mut containers: ContainerCollection,
    gen_settings: impl Borrow<GenerationSettings>,
    generate_package_name: String,
) -> Exports<Python> {
    let generate_package_name = generate_package_name.to_case(Case::Snake);
    let gen_settings = gen_settings.borrow();

    if !gen_settings.module_structure {
        containers.flatten();
    }

    let mut files = Vec::new();

    files.push(ExportFile {
        content_type: "util".to_owned(),
        content: gen_util(),
    });

    files.push(ExportFile {
        content_type: "basic_types".to_owned(),
        content: gen_basic_typings(),
    });

    files.extend(gen_typings(&containers, generate_package_name.clone()));

    if gen_settings.runtime_type_checks {
        let type_checks = gen_type_checks(containers.all_containers());

        let type_checks = quote! {
            from .util import *
            from .types import *

            $type_checks
        };

        files.push(ExportFile {
            content_type: "type_checks".to_owned(),
            content: type_checks,
        });
    }

    if gen_settings.ser {
        let serializer_code = gen_serializer_code();
        let ser_code = quote! {
            from typing import Union

            from .types import *
            from .util import *
            from .serializer import Serializer

            $(gen_ser_functions(containers.all_containers()))

            $(gen_serialize_func(containers.all_containers(), gen_settings.runtime_type_checks))
        };

        files.push(ExportFile {
            content_type: "serializer".to_owned(),
            content: serializer_code,
        });

        files.push(ExportFile {
            content_type: "ser".to_owned(),
            content: ser_code,
        });
    }

    if gen_settings.des {
        let deserializer_code = gen_deserializer_code();
        let des_code = quote! {
            from typing import TypeVar, Type, cast

            from .types import *
            from .util import *
            from .deserializer import Deserializer

            $(gen_des_functions(containers.all_containers()))

            $(gen_deserialize_func(containers.all_containers()))
        };

        files.push(ExportFile {
            content_type: "deserializer".to_owned(),
            content: deserializer_code,
        });

        files.push(ExportFile {
            content_type: "des".to_owned(),
            content: des_code,
        });
    }

    let mut import_registry = ImportRegistry::new(generate_package_name);
    import_registry.push(Package::Relative("types".into()), ImportItem::All);
    import_registry.push(Package::Relative("basic_types".into()), ImportItem::All);

    if gen_settings.des {
        import_registry.push(
            Package::Relative("des".into()),
            ImportItem::Single("deserialize".into()),
        );
    }

    if gen_settings.ser {
        import_registry.push(
            Package::Relative("ser".into()),
            ImportItem::Single("serialize".into()),
        );
    }

    files.push(ExportFile {
        content_type: "__init__".to_owned(),
        content: quote!($import_registry),
    });

    Exports { files }
}

impl<I> TokensIterExt<Python> for I
where
    I: Iterator<Item = Tokens>,
{
    const LOGICAL_AND: &'static str = PYTHON_LOGIC_AND;
    const LOGICAL_OR: &'static str = PYTHON_LOGIC_OR;
}

pub(super) struct BranchedTemplate;

impl IfBranchedTemplate<Python> for BranchedTemplate {
    const IF_BRANCH: &'static str = "if";
    const IF_ELSE_BRANCH: &'static str = "elif";
    const ELSE_BRANCH: &'static str = "else";

    fn push_condition(tokens: &mut Tokens, condition: impl FormatInto<Python>) {
        tokens.append(condition)
    }

    fn push_condition_block(tokens: &mut Tokens, body: impl FormatInto<Python>) {
        tokens.append(":");
        tokens.indent();
        tokens.append(body);
        tokens.unindent();
    }
}

impl<I> TokensBranchedIterExt<Python> for I
where
    I: Iterator<Item = (Option<Tokens>, Tokens)>,
{
    type Template = BranchedTemplate;
}

impl FormatInto<Python> for FieldAccessor<'_> {
    fn format_into(self, tokens: &mut Tokens) {
        quote_in! { *tokens =>
            $(match self {
                Self::Array | Self::None => (),
                Self::Object(n) => $n = $[' '],
            })
        }
    }
}

impl FormatInto<Python> for VariablePath {
    fn format_into(self, tokens: &mut genco::Tokens<Python>) {
        quote_in! { *tokens =>
            $(self.start_variable)
        }
        self.parts
            .into_iter()
            .for_each(|part| part.format_into(tokens))
    }
}

impl Default for VariablePath {
    fn default() -> Self {
        Self::new(PYTHON_OBJECT_VARIABLE.to_owned())
    }
}

impl FormatInto<Python> for VariableAccess {
    fn format_into(self, tokens: &mut genco::Tokens<Python>) {
        quote_in! { *tokens =>
            $(match self {
                Self::Indexed(index) => [$index],
                Self::Field(name) => .$name,
            })
        }
    }
}

impl FormatInto<Python> for AvailableCheck {
    fn format_into(self, tokens: &mut Tokens) {
        quote_in! { *tokens =>
            $(match self {
                AvailableCheck::Object(..) => (),
                AvailableCheck::None => ()
            })
        }
    }
}

impl FormatInto<Python> for ImportRegistry {
    fn format_into(self, tokens: &mut Tokens) {
        let (base_path, items) = self.into_items_sorted();
        for (package, imports) in items {
            let joiner = ".";
            let package = match package {
                Package::Relative(path) => format!(".{}", path.into_path(joiner)),
                Package::Extern(path) => path.into_path(joiner).to_string(),
                Package::Intern(mut path) => {
                    if !path.is_empty() {
                        path.push_front(base_path.as_str());
                        path.into_path(joiner).to_string()
                    } else {
                        PathBuf::new()
                            .join(base_path.as_str())
                            .into_path(joiner)
                            .to_string()
                    }
                }
            };

            quote_in!(*tokens=> from $(package) import);
            tokens.space();

            match imports {
                ImportMode::All => quote_in!(*tokens=> *),
                ImportMode::Single(items) => {
                    let items = items.iter().map(|i| {
                        if let Some(alias) = &i.alias {
                            quote!($(&i.name) as $alias)
                        } else {
                            quote!($(&i.name))
                        }
                    });
                    quote_in!(*tokens=> $(for part in items join (, ) => $part))
                }
            }

            tokens.push();
        }
    }
}

#[cfg(test)]
mod test {
    use genco::tokens::FormatInto;

    use super::Tokens;

    #[test]
    fn test_import_registry_format() {
        use super::{ImportItem, ImportRegistry, Package};

        let mut import_registry = ImportRegistry::new("package".to_owned());
        import_registry.push(
            Package::Relative("basic_types".into()),
            ImportItem::Aliased {
                item_name: "A".into(),
                alias: "A__A".into(),
            },
        );
        import_registry.push(
            Package::Intern("des".into()),
            ImportItem::Single("deserialize".into()),
        );
        import_registry.push(
            Package::Extern("ser".into()),
            ImportItem::Single("serialize".into()),
        );
        import_registry.push(Package::Relative("types".into()), ImportItem::All);

        let mut tokens = Tokens::new();
        import_registry.format_into(&mut tokens);

        assert_eq!(
            tokens.to_file_string().unwrap(),
            format!(
                r#"from ser import serialize
from package.des import deserialize
from .basic_types import A as A__A
from .types import *
"#
            )
        );
    }
}
