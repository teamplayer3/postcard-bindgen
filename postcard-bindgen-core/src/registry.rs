use core::fmt::Display;
use std::{borrow::Cow, collections::VecDeque};

use alloc::vec::Vec;
use tree_ds::prelude::{Node, NodeRemovalStrategy, Tree};

use crate::{
    path::Path,
    type_info::{GenJsBinding, ValueType},
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Container {
    pub path: Path<'static, 'static>,
    pub name: &'static str,
    pub r#type: BindingType,
}

impl Container {
    pub fn flatten_paths(&mut self) {
        self.path.flatten();

        match self.r#type {
            BindingType::Struct(ref mut ty) => ty.flatten_paths(),
            BindingType::Enum(ref mut ty) => ty.flatten_paths(),
            BindingType::TupleStruct(ref mut ty) => ty.flatten_paths(),
            _ => (),
        }
    }
}

pub struct ContainerInfo<'a> {
    pub name: Cow<'a, str>,
    pub path: Path<'a, 'a>,
}

impl From<&Container> for ContainerInfo<'_> {
    fn from(container: &Container) -> Self {
        Self {
            name: Cow::Borrowed(container.name),
            path: container.path.clone(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum BindingType {
    Struct(StructType),
    TupleStruct(TupleStructType),
    UnitStruct(UnitStructType),
    Enum(EnumType),
}

#[derive(Debug, Default, Clone, PartialEq, Eq)]
// encoded into | variant index | (inner)
pub struct EnumType {
    pub variants: Vec<EnumVariant>,
}

impl EnumType {
    pub fn new() -> Self {
        Self::default()
    }

    // index is set based on order of variant registration
    pub fn register_variant(&mut self, name: &'static str) {
        self.variants.push(EnumVariant {
            index: self.variants.len(),
            name,
            inner_type: EnumVariantType::Empty,
        });
    }

    pub fn register_variant_tuple(&mut self, name: &'static str, fields: TupleFields) {
        self.variants.push(EnumVariant {
            index: self.variants.len(),
            name,
            inner_type: EnumVariantType::Tuple(fields.into_inner()),
        });
    }

    pub fn register_unnamed_struct(&mut self, name: &'static str, fields: StructFields) {
        self.variants.push(EnumVariant {
            index: self.variants.len(),
            name,
            inner_type: EnumVariantType::NewType(fields.into_inner()),
        })
    }

    fn flatten_paths(&mut self) {
        for variant in &mut self.variants {
            match &mut variant.inner_type {
                EnumVariantType::NewType(fields) => {
                    for field in fields {
                        field.v_type.flatten_paths();
                    }
                }
                EnumVariantType::Tuple(fields) => {
                    for field in fields {
                        field.flatten_paths();
                    }
                }
                _ => {}
            }
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct EnumVariant {
    pub index: usize,
    pub name: &'static str,
    pub inner_type: EnumVariantType,
}

impl AsRef<EnumVariant> for EnumVariant {
    fn as_ref(&self) -> &EnumVariant {
        self
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum EnumVariantType {
    Empty,
    Tuple(Vec<ValueType>),
    // for unnamed structs create struct with custom name ( __EnumName_Struct1)
    NewType(Vec<StructField>),
}

#[derive(Debug, Default, Clone, PartialEq, Eq)]
pub struct StructType {
    pub fields: Vec<StructField>,
}

impl StructType {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn register_field<T: GenJsBinding>(&mut self, name: &'static str) {
        self.fields.push(StructField {
            name,
            v_type: T::get_type(),
        })
    }

    fn flatten_paths(&mut self) {
        for field in &mut self.fields {
            field.v_type.flatten_paths();
        }
    }
}

#[derive(Debug, Default, Clone, PartialEq, Eq)]
pub struct TupleStructType {
    pub fields: Vec<ValueType>,
}

impl TupleStructType {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn register_field<T: GenJsBinding>(&mut self) {
        self.fields.push(T::get_type())
    }

    fn flatten_paths(&mut self) {
        for field in &mut self.fields {
            field.flatten_paths();
        }
    }
}

#[derive(Debug, Default, Clone, PartialEq, Eq)]
pub struct UnitStructType;

impl UnitStructType {
    pub fn new() -> Self {
        Self {}
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StructField {
    pub name: &'static str,
    pub v_type: ValueType,
}

#[derive(Debug, Default)]
pub struct StructFields(Vec<StructField>);

impl StructFields {
    pub fn register_field<T: GenJsBinding>(&mut self, name: &'static str) {
        self.0.push(StructField {
            name,
            v_type: T::get_type(),
        })
    }

    fn into_inner(self) -> Vec<StructField> {
        self.0
    }
}

#[derive(Default)]
pub struct TupleFields(Vec<ValueType>);

impl TupleFields {
    pub fn register_field<T: GenJsBinding>(&mut self) {
        self.0.push(T::get_type())
    }

    fn into_inner(self) -> Vec<ValueType> {
        self.0
    }
}

pub struct ContainerCollection(Tree<NodeId, NodeType>);

impl ContainerCollection {
    pub fn flatten(&mut self) {
        let root_node_id = self.0.get_root_node().unwrap().get_node_id();
        let root_node = self.0.get_node_by_id(&root_node_id).unwrap();

        let root_mod_nodes = root_node
            .get_children_ids()
            .iter()
            .filter_map(|i| {
                let node = self.0.get_node_by_id(i).unwrap();

                let node_value = node.get_value().unwrap();
                if node_value.is_module() {
                    Some(node)
                } else {
                    None
                }
            })
            .collect::<VecDeque<_>>();

        let mut mod_nodes = root_mod_nodes.clone();

        loop {
            if mod_nodes.is_empty() {
                break;
            }

            let node = mod_nodes.pop_front().unwrap();

            let children = node.get_children_ids();

            for child in children {
                let child_node = self.0.get_node_by_id(&child).unwrap();
                let child_value = child_node.get_value().unwrap();

                node.remove_child(child_node.clone());

                if child_value.is_container() {
                    root_node.add_child(child_node);
                } else if child_value.is_module() {
                    mod_nodes.push_back(child_node);
                }
            }
        }

        for node in root_mod_nodes {
            self.0
                .remove_node(
                    &node.get_node_id(),
                    NodeRemovalStrategy::RemoveNodeAndChildren,
                )
                .unwrap();
        }

        for node in root_node
            .get_children_ids()
            .iter()
            .map(|node| self.0.get_node_by_id(node).unwrap())
        {
            node.update_value(|v| v.as_mut().unwrap().container_mut().unwrap().flatten_paths());
        }
    }

    pub fn all_containers(&self) -> impl Iterator<Item = Container> + Clone + '_ {
        self.0
            .get_nodes()
            .iter()
            .filter_map(|node| node.get_value().unwrap().container().cloned())
    }

    pub fn containers_per_module(&self) -> (Vec<Container>, Vec<Module<'_>>) {
        let root_node = self.0.get_root_node().unwrap().get_node_id();
        container_and_modules_per_mod(&self.0, &root_node)
    }
}

#[derive(Debug, Clone)]
pub struct Module<'a> {
    tree: &'a Tree<NodeId, NodeType>,
    node_id: NodeId,
    name: Cow<'static, str>,
    cached_path: Option<String>,
}

impl<'a> Module<'a> {
    fn new(tree: &'a Tree<NodeId, NodeType>, node_id: NodeId, name: Cow<'static, str>) -> Self {
        Self {
            tree,
            node_id,
            name,
            cached_path: None,
        }
    }

    pub fn path(&self) -> String {
        if let Some(path) = &self.cached_path {
            path.clone()
        } else {
            let mut curr_node = self.tree.get_node_by_id(&self.node_id).unwrap();
            let mut path = Vec::new();
            while let Some(id) = curr_node.get_parent_id() {
                let node = self.tree.get_node_by_id(&id).unwrap();
                if let NodeType::Module(name) = node.get_value().unwrap() {
                    if name != "::" {
                        path.push(name);
                        curr_node = node;
                    } else {
                        break;
                    }
                }
            }

            path.join("/")
        }
    }

    pub fn name(&self) -> &str {
        self.name.as_ref()
    }

    pub fn entries(&self) -> (Vec<Container>, Vec<Module<'a>>) {
        container_and_modules_per_mod(self.tree, &self.node_id)
    }
}

fn container_and_modules_per_mod<'a>(
    tree: &'a Tree<NodeId, NodeType>,
    node_id: &NodeId,
) -> (Vec<Container>, Vec<Module<'a>>) {
    let node = tree.get_node_by_id(node_id).unwrap();

    node.sort_children(|a, b| {
        let a_height = tree.get_node_height(a).unwrap();
        let b_height = tree.get_node_height(b).unwrap();

        a_height.cmp(&b_height).reverse()
    });

    let mut mods = Vec::new();
    let mut containers = Vec::new();

    for (id, child) in node
        .get_children_ids()
        .iter()
        .map(|id| (id, tree.get_node_by_id(id).unwrap()))
    {
        match child.get_value().unwrap() {
            NodeType::Module(name) => mods.push(Module::new(tree, *id, name)),
            NodeType::Container(container) => containers.push(container.clone()),
        }
    }

    (containers, mods)
}

type NodeId = u128;

enum PathExists {
    Full(NodeId),
    Partly(NodeId, Cow<'static, str>),
}

#[derive(Debug, Clone, PartialEq, Eq)]
enum NodeType {
    Container(Container),
    Module(Cow<'static, str>),
}

impl Default for NodeType {
    fn default() -> Self {
        NodeType::Module("_".into())
    }
}

impl Display for NodeType {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            NodeType::Container(c) => write!(f, "Container({}, {})", c.name, c.path),
            NodeType::Module(m) => write!(f, "Module({})", m),
        }
    }
}

impl NodeType {
    fn container(&self) -> Option<&Container> {
        match self {
            NodeType::Container(c) => Some(c),
            _ => None,
        }
    }

    fn container_mut(&mut self) -> Option<&mut Container> {
        match self {
            NodeType::Container(c) => Some(c),
            _ => None,
        }
    }

    fn is_module(&self) -> bool {
        matches!(self, NodeType::Module(_))
    }

    fn is_container(&self) -> bool {
        matches!(self, NodeType::Container(_))
    }
}

#[derive(Debug)]
pub struct BindingsRegistry(Tree<NodeId, NodeType>);

impl BindingsRegistry {
    pub fn register_struct_binding(
        &mut self,
        name: &'static str,
        path: impl Into<Cow<'static, str>>,
        value: StructType,
    ) {
        self.insert_container(Container {
            path: Path::new(path, "::"),
            name,
            r#type: BindingType::Struct(value),
        });
    }

    pub fn register_tuple_struct_binding(
        &mut self,
        name: &'static str,
        path: impl Into<Cow<'static, str>>,
        value: TupleStructType,
    ) {
        self.insert_container(Container {
            path: Path::new(path, "::"),
            name,
            r#type: BindingType::TupleStruct(value),
        });
    }

    pub fn register_unit_struct_binding(
        &mut self,
        name: &'static str,
        path: impl Into<Cow<'static, str>>,
        value: UnitStructType,
    ) {
        self.insert_container(Container {
            path: Path::new(path, "::"),
            name,
            r#type: BindingType::UnitStruct(value),
        });
    }

    pub fn register_enum_binding(
        &mut self,
        name: &'static str,
        path: impl Into<Cow<'static, str>>,
        value: EnumType,
    ) {
        self.insert_container(Container {
            path: Path::new(path, "::"),
            name,
            r#type: BindingType::Enum(value),
        });
    }

    pub fn into_entries(self) -> ContainerCollection {
        ContainerCollection(self.0)
    }

    fn insert_container(&mut self, container: Container) {
        let mut node = self.0.get_root_node().unwrap();
        let node_id = {
            let container_path = &container.path;
            let mut parts = container_path
                .parts()
                .skip(1)
                .map(|p| p.to_owned())
                .peekable();
            let path_exists = loop {
                let part = parts.next();
                let is_last = parts.peek().is_none();

                if let Some(part) = part {
                    let node_ids = node.get_children_ids();
                    let child = node_ids.iter().find(|child| {
                        let node = self.0.get_node_by_id(child).unwrap();
                        matches!(node.get_value(), Some(NodeType::Module(p)) if p == part)
                    });

                    if let Some(child) = child {
                        if is_last {
                            break PathExists::Full(*child);
                        }

                        node = self.0.get_node_by_id(child).unwrap();
                    } else {
                        break PathExists::Partly(node.get_node_id(), part.into());
                    }
                } else {
                    break PathExists::Full(node.get_node_id());
                }
            };

            match path_exists {
                PathExists::Full(node_id) => node_id,
                PathExists::Partly(node_id, part) => {
                    let mut node = self
                        .0
                        .add_node(
                            Node::new_with_auto_id(Some(NodeType::Module(part.clone()))),
                            Some(&node_id),
                        )
                        .unwrap();

                    for part in parts {
                        node = self
                            .0
                            .add_node(
                                Node::new_with_auto_id(Some(NodeType::Module(part.clone().into()))),
                                Some(&node),
                            )
                            .unwrap();
                    }

                    node
                }
            }
        };

        self.0
            .add_node(
                Node::new_with_auto_id(Some(NodeType::Container(container))),
                Some(&node_id),
            )
            .unwrap();
    }
}

impl Default for BindingsRegistry {
    fn default() -> Self {
        let mut tree: Tree<NodeId, NodeType> = Tree::new(None);
        tree.add_node(
            Node::new_with_auto_id(Some(NodeType::Module("::".into()))),
            None,
        )
        .unwrap();
        Self(tree)
    }
}

pub trait JsBindings {
    fn create_bindings(registry: &mut BindingsRegistry);
}

#[cfg(test)]
mod test {
    use crate::registry::{
        BindingsRegistry, EnumType, JsBindings, StructFields, StructType, TupleFields,
        TupleStructType,
    };

    #[test]
    fn test_registry_struct() {
        #[allow(unused)]
        struct Test {
            a: u8,
            b: u16,
            c: &'static str,
        }

        impl JsBindings for Test {
            fn create_bindings(registry: &mut BindingsRegistry) {
                let mut ty = StructType::new();

                ty.register_field::<u8>("a");
                ty.register_field::<u16>("b");
                ty.register_field::<&str>("c");

                registry.register_struct_binding("Test", "", ty);
            }
        }

        let mut registry = BindingsRegistry::default();
        Test::create_bindings(&mut registry);
    }

    #[test]
    fn test_registry_tuple_struct() {
        #[allow(dead_code)]
        struct Test(u8, &'static str, &'static [u8]);

        impl JsBindings for Test {
            fn create_bindings(registry: &mut BindingsRegistry) {
                let mut ty = TupleStructType::new();

                ty.register_field::<u8>();
                ty.register_field::<&str>();
                ty.register_field::<&[u8]>();

                registry.register_tuple_struct_binding("Test", "", ty);
            }
        }

        let mut registry = BindingsRegistry::default();
        Test::create_bindings(&mut registry);
    }

    #[test]
    fn test_registry_enum() {
        #[allow(unused)]
        enum Test {
            A,
            B(u8),
            C { a: &'static str, b: u16 },
        }

        impl JsBindings for Test {
            fn create_bindings(registry: &mut BindingsRegistry) {
                let mut ty = EnumType::new();

                ty.register_variant("A");

                let mut fields = TupleFields::default();
                fields.register_field::<u8>();
                ty.register_variant_tuple("B", fields);

                let mut fields = StructFields::default();
                fields.register_field::<&str>("a");
                fields.register_field::<u16>("b");
                ty.register_unnamed_struct("C", fields);

                registry.register_enum_binding("Test", "", ty);
            }
        }

        let mut registry = BindingsRegistry::default();
        Test::create_bindings(&mut registry);
    }
}
