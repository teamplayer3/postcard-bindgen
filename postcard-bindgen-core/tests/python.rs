mod registry;

use postcard_bindgen_core::code_gen::python::{generate, GenerationSettings};

use registry::init_registry;

#[test]
fn test_runtime_checks() {
    let containers = init_registry().into_entries();

    let gen_settings = GenerationSettings::enable_all();
    let exports = generate(containers, gen_settings, "test".to_owned());

    let runtime_checks_file = exports
        .file("runtime_checks")
        .unwrap()
        .to_file_string()
        .unwrap();
    insta::assert_snapshot!(runtime_checks_file);
}

#[test]
fn test_ser_with_runtime_checks() {
    let containers = init_registry().into_entries();

    let gen_settings = GenerationSettings::enable_all();
    let exports = generate(containers, gen_settings, "test".to_owned());

    let ser_file = exports.file("ser").unwrap().to_file_string().unwrap();
    insta::assert_snapshot!(ser_file);
}

#[test]
fn test_ser_without_runtime_checks() {
    let containers = init_registry().into_entries();

    let gen_settings = GenerationSettings::enable_all().runtime_type_checks(false);
    let exports = generate(containers, gen_settings, "test".to_owned());

    let ser_file = exports.file("ser").unwrap().to_file_string().unwrap();
    insta::assert_snapshot!(ser_file);
}

#[test]
fn test_des() {
    let containers = init_registry().into_entries();

    let gen_settings = GenerationSettings::enable_all();
    let exports = generate(containers, gen_settings, "test".to_owned());

    let des_file = exports.file("des").unwrap().to_file_string().unwrap();
    insta::assert_snapshot!(des_file);
}
