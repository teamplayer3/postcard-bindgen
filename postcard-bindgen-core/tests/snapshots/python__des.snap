---
source: postcard-bindgen-core/tests/python.rs
expression: des_file
snapshot_kind: text
---
from typing import TypeVar, Type, cast

from .types import *
from .util import *
from .deserializer import Deserializer

def deserialize_STRUCT_TYPE(d) -> StructType:
    return StructType(field_1 = d.deserialize_number(U32_BYTES, False), field_2 = d.deserialize_string(), field_3 = d.deserialize_array(lambda d: d.deserialize_number(U32_BYTES, False), None), field_4 = d.deserialize_array(lambda d: deserialize_STRUCT_TYPE(d), None), field_5 = range(d.deserialize_number(U32_BYTES, False), d.deserialize_number(U32_BYTES, False)), field_6 = d.deserialize_map((lambda d: (d.deserialize_string(), d.deserialize_number(U32_BYTES, False)))), field_7 = d.deserialize_map((lambda d: (d.deserialize_number(U32_BYTES, False), d.deserialize_number(U32_BYTES, False)))), field_8 = None if d.deserialize_number(U32_BYTES, False) == 0 else d.deserialize_number(U32_BYTES, False), field_9 = d.deserialize_bool(), field_10 = (d.deserialize_number(U32_BYTES, False), d.deserialize_string()), field_11 = d.deserialize_array(lambda d: d.deserialize_string(), 3), field_12 = d.deserialize_array(lambda d: d.deserialize_number(U32_BYTES, False), None), field_13 = d.deserialize_number_float(U32_BYTES))

def deserialize_UNIT_STRUCT_TYPE(d) -> UnitStructType:
    return UnitStructType()

def deserialize_TUPLE_STRUCT_TYPE(d) -> TupleStructType:
    return TupleStructType(d.deserialize_number(U32_BYTES, False), d.deserialize_string())

def deserialize_ENUM_TYPE(d) -> EnumType:
    variant_index = d.deserialize_number(U32_BYTES, False)
    if variant_index == 0:
        return EnumType_AVariant()
    elif variant_index == 1:
        return EnumType_BVariant(d.deserialize_number(U32_BYTES, False), d.deserialize_string())
    elif variant_index == 2:
        return EnumType_CVariant(field_1 = d.deserialize_number(U32_BYTES, False), field_2 = d.deserialize_string(), struct_type = deserialize_STRUCT_TYPE(d))
    else:
        raise TypeError("variant index {} not exists".format(variant_index))

T = TypeVar("T", StructType, UnitStructType, TupleStructType, EnumType)
def deserialize(obj_type: Type[T], bytes: bytes) -> T:
    d = Deserializer(bytes)

    if obj_type is StructType:
        return cast(T, deserialize_STRUCT_TYPE(d))
    elif obj_type is UnitStructType:
        return cast(T, deserialize_UNIT_STRUCT_TYPE(d))
    elif obj_type is TupleStructType:
        return cast(T, deserialize_TUPLE_STRUCT_TYPE(d))
    elif obj_type is EnumType:
        return cast(T, deserialize_ENUM_TYPE(d))
    else:
        raise TypeError("{} not deserializable".format(obj_type))
